[futures] Add collectAny() to ivalue::Future for completeness

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39597 [futures] Add collectAny() to ivalue::Future for completeness**

To complement collectAll(), this change adds collectAny(), and writes
up relevant unittest coverage.

We also remove the vector-based helper version of collectAll(), which
was debatable usefulness in a previous change.

Differential Revision: [D21910311](https://our.internmc.facebook.com/intern/diff/D21910311/)

Commit:
[futures] Add collectAny() to ivalue::Future for completeness

To complement collectAll(), this change adds collectAny(), and writes
up relevant unittest coverage.

We also remove the vector-based helper version of collectAll(), which
was debatable usefulness in a previous change.

Differential Revision: [D21910311](https://our.internmc.facebook.com/intern/diff/D21910311/)

[ghstack-poisoned]

wanchaol commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -634,21 +634,47 @@ CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
-  auto typePtr = srcs.empty() ? AnyType::get() : srcs[0]->elementType();
-  // Lists are generally expected to have a consistent type, check
-  // for this unless we have a compelling target use case.
-  for (size_t i = 1, len = srcs.size(); i < len; ++i) {
```
I think we should still those torch check logics in the `collectAll` c10::List overload, to ensure the single type list constraint.

jjlilley replied:
Sounds good,  adding a `TORCH_CHECK` to the `addCallback` loop.

wanchaol commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -634,21 +634,47 @@ CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
+    if (srcs.get(i)->completed()) {
+      return srcs.get(i);
+    }
+    TORCH_CHECK(i == 0 || (*typePtr == *srcs.get(i)->elementType()));
```
nit: we don't need i==0 here?

jjlilley replied:
The `i==0` is to skip the (more expensive) type check on the right for the i ==0 case

wanchaol reviewed (approved):
looks good except some minor comments.

Commit:
add TORCH_CHECK, per review comment on "[futures] Add collectAny() to ivalue::Future for completeness"


To complement collectAll(), this change adds collectAny(), and writes
up relevant unittest coverage.

We also remove the vector-based helper version of collectAll(), which
was debatable usefulness in a previous change.

Differential Revision: [D21910311](https://our.internmc.facebook.com/intern/diff/D21910311/)

[ghstack-poisoned]

Commit:
ran clang-format on test_misc.cpp on "[futures] Add collectAny() to ivalue::Future for completeness"


To complement collectAll(), this change adds collectAny(), and writes
up relevant unittest coverage.

We also remove the vector-based helper version of collectAll(), which
was debatable usefulness in a previous change.

Differential Revision: [D21910311](https://our.internmc.facebook.com/intern/diff/D21910311/)

[ghstack-poisoned]
